go-controller/pkg/ovn/libovsdbops/util.go

```diff
@@ -2,38 +2,396 @@ package libovsdbops
+				}
+			} else {
+				ops = append(ops, o...)
+			}
+		} else if opModel.OnModelMutations != nil {
+			o, err := m.mutate(true, opModel.Model, opModel)
+			if err != nil {
+				if errors.Is(err, client.ErrNotFound) {
+					o, err = m.create(opModel)
+					if err != nil {
+						return nil, err
+					}
+					ops = append(ops, o...)
+				} else {
+					return nil, err
+				}
+			} else {
+				ops = append(ops, o...)
+			}
+		} else {
+			if _, err := m.get(opModel.Model); errors.Is(err, client.ErrNotFound) {
+				o, err := m.create(opModel)
+				if err != nil {
+					return nil, err
+				}
+				ops = append(ops, o...)
+			}
+		}
+	}
+	return ops, nil
+}
+
+/*
+ Delete performs idempotent delete operations against libovsdb according to the
+ following logic:
+
+ a) if a ModelPredicate is provided; it uses a lookup of the Model in the cache,
+ following the predicate function specified.
+
+ b) if a) is not true, but OnModelMutations is specified; it performs a direct
+ mutation of the Model if it exists
+
+ c) if a) and b) are not true, but Model is specified; it performs a direct
+ delete of the Model.
+
+ This function does not assume that only one object is retrieved for a given
+ predicate, a delete/mutate operation will be performed on all retrived models
+ should the predicate return multiple.
+*/
+func (m *ModelClient) Delete(opModels ...OperationModel) error {
+	ops, err := m.DeleteOps(opModels...)
+	if err != nil {
+		return err
+	}
+	_, err = TransactAndCheck(m.client, ops)
+	return err
+}
+
+func (m *ModelClient) DeleteOps(opModels ...OperationModel) (ops []ovsdb.Operation, err error) {
+	for _, opModel := range opModels {
+		if opModel.ModelPredicate != nil {
+			if err := m.client.WhereCache(opModel.ModelPredicate).List(opModel.ExistingResult); err != nil {
+				return nil, fmt.Errorf("unable to list items for model, err: %v", err)
+			}
+			for i := 0; i < reflect.ValueOf(opModel.ExistingResult).Elem().Len(); i++ {
+				if opModel.OnModelMutations != nil {
+					o, err := m.mutate(false, reflect.ValueOf(opModel.ExistingResult).Elem().Index(i).Addr().Interface(), opModel)
+					if err != nil {
+						return nil, err
+					}
+					ops = append(ops, o...)
+				} else {
+					o, err := m.delete(false, reflect.ValueOf(opModel.ExistingResult).Elem().Index(i).Addr().Interface(), opModel)
+					if err != nil {
+						return nil, err
+					}
+					ops = append(ops, o...)
+				}
+			}
+		} else if opModel.OnModelMutations != nil {
+			o, err := m.mutate(true, opModel.Model, opModel)
+			if err != nil {
+				return nil, err
+			}
+			ops = append(ops, o...)
+		} else if opModel.Model != nil {
+			o, err := m.delete(true, opModel.Model, opModel)
+			if err != nil {
+				return nil, err
+			}
+			ops = append(ops, o...)
+		}
+	}
+	return ops, nil
+}
+
+func (m *ModelClient) createOrUpdateWithCache(opModel OperationModel) ([]ovsdb.Operation, error) {
+	if err := m.client.WhereCache(opModel.ModelPredicate).List(opModel.ExistingResult); err != nil {
```

Comment:
In line with my comment above, right now to do lookups we have to specify an `OperationModel` with `ModelPredicate` and use `WhereCache` when there are situations where we could use a more efficient `Get` instead.

My suggestion for `createOrUpdate` , which is not perfect, but I think is the best thing to do currently, is:
- if `ExistingResult` was provided: 
  - If `opModel.ModelPredicate` is provided, do `WhereCache` and add result to `ExistingResult`.
  - Else, do `Get` on `opModel.Model` and add result to `ExistingResult`.
- If `ExistingResult` was not provided or is empty, do `create`
- Otherwise proceed with update/mutation using the lookup that was already done previously.

I think this also leads to simpler code.